### ISSUE TITLE
Lens flare effects

### ISSUE THREAD

**Issue by hrydgard** [hrydgard/ppsspp#15923]
Lens flares are a typical problematic effect on GPUs of the PSP's generation. They are supposed to be drawn only when the sun (or other light source) is visible, but there are no occlusion queries you can use to figure out if it is directly on the GPU, neither is it practical to copy the texel to an image and then use multitexturing to blend the lens effect texture with the copied texel, since multitexturing is not a thing.

So games make use of a variety of dirty tricks.

Let's start with Wipeout Pure, #13344. I started by hacking the interpreter to log out CPU reads from VRAM. For some reason there are a whole bunch that happen every frame, but these stand out: 

```
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471bdec
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471bdf0
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471bdf4
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471bdf8

06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471c5ec
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471c5f0
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471c5f4
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471c5f8

06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471cdec
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471cdf0
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471cdf4
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471cdf8

06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471d5ec
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471d5f0
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471d5f4
06:50:095 idle0        W[G3D]: MIPS\MIPSInt.cpp:411 Read from depth buffer? addr=0471d5f8
```
It's using lh instructions (load 16-bit) and it looks to me like it's sampling a 4x4 rectangle around the sun's position from the depth buffer, skipping every other pixel - it is situated at 110000 in VRAM which starts at 04000000, plus the 600000 deswizzle offset that's needed to linearize the depth buffer in 8888 mode. As expected, as the sun slides across the image when the camera moves, these addresses, which are read from every frame, change accordingly.

For this to work correctly, we have to read back the depth buffer every frame to emulated PSP VRAM, which introduces a massive sync point between the GPU and CPU. This is not really desirable (although we should implement it as an option), so I've been thinking about ways to get around it:

* Unsynchronized readbacks (Vulkan)
  - Schedule readbacks on the GPU's timeline, but don't wait for them to complete, instead have a background thread wait on a fence from the GPU, copying to the PSP's VRAM when done, whenever that is, in the background. For the purposes of lens flare visibility, this might be fine. No CPU stall, but the readback might be delayed, which could in theory (and given our luck, probably in practice too) corrupt important data during level transitions and similar.
* Virtual readbacks with hooks 
  - As above, but we don't actually read back to the PSP's VRAM but to an alternative buffer, then for each affected game, hook the PSP function that reads from the depth buffer to read the data from the alternative buffer instead. This has the advantage that we don't even need to copy the whole depth buffer from GPU memory to PSP RAM, instead we can read directly from there (need to be careful about memory types for performance), but requires game specific work.

.... To be continued

**Comment by Gamemulatorer** [comment]
Artic Edge
https://github.com/hrydgard/ppsspp/issues/11100#issuecomment-1123170383
NFS Shift
https://github.com/hrydgard/ppsspp/issues/11100#issuecomment-1123377372

**Comment by hrydgard** [comment]
Thanks, added to the list.

**Comment by QrTa** [comment]
Syphon Filter Logan's Shadow
https://github.com/hrydgard/ppsspp/issues/10229#issuecomment-1232151181

**Comment by 71knight** [comment]
I don't know how aethersx2 PS2 emulator does it, but they get accurate full speed readback emulation using opengl on Android.  I can play need for speed hot pursuit 2 without underclocking the emulator and have accurate readbacks turned on and maintain full speed emulation with no slowdowns and the sun hides when it's supposed to.   And I think the PS2 has double the resolution of PSP.  I do cheat a little though... I keep all my cores frequencies maxed out and GPU set at 3/4 speed on my rooted phone.  The phone doesn't get too hot.... About 147° F on average.  So I know ppsspp wouldn't be too demanding with accurate readbacks.  I think what helps them is they have CPU affinity option that keeps the heaviest threads on the biggest cores of the phone.

**Comment by Gamemulatorer** [comment]
Aethersx2 is the pcsx2 mobile port btw.

**Comment by hrydgard** [comment]
An alternative to readbacks for the games that peek the Z-buffer using the CPU, as commented elsewhere by @unknownbrackets , would be to run both the software and hardware renderers side by side, that way we'll always have accurate depth in CPU-accessible memory, at the right time. 

This is expensive though, and to make it less so, it would be possible to have the software renderer only render depth buffers, and just ignore color - depth is a lot less complex so I think this would be way faster than running the full software renderer. This wouldn't work for cases where games reinterpret color and Z like Kuroyou, but I don't think that applies to any of these cases.

Also gonna have to look into what PCSX2 does. Maybe SX2 Aether does something special on top, hard to say given it's close source..

**Comment by unknownbrackets** [comment]
I will say, the loop to interpolate triangle data is the slowest part of the software renderer now, I think.  Texture sampling is still fairly slow as well.

We'd still need to texture (because of alpha tests/color tests), but we could skip alpha blend and logicops.  Skipping blending would save time, but I don't think it'd make a huge difference overall.

Maybe we could have a "fast and loose" mode where it ignores color and alpha tests, though, or at least skip sampling/etc. when they're not enabled.  That would also allow us to skip lighting which is quite expensive.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah I think we can go very fast and loose for Z-only. Texturing only needs to be done when we know there's alpha. And we could skip filtering and mipmapping for example..

**Comment by unknownbrackets** [comment]
Right.  My biggest concern would be "depth boxes" from alpha testing.  For example, if some trees or clouds were drawn to cover the sun, but without alpha testing they cover the entire thing.  If we can safely skip alpha testing, it probably helps the potential speed a lot, because it cuts out many, many things.

We might end up in a place where we're using heuristics to skip alpha testing, though.  For example, it's probably mainly an issue with flat Z - models probably don't need alpha testing for depth to be correct.

-[Unknown]

**Comment by Gamemulatorer** [comment]
Socom US Navy Seal: Tactical Strike is also affected.
https://github.com/hrydgard/ppsspp/issues/15071
![Screenshot_2022-10-13-02-23-08-77](https://user-images.githubusercontent.com/37603562/195419965-baa91c32-be33-4b3a-acda-559747824e4a.png)

[UCUS98649.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/9767756/UCUS98649.ppdmp.zip)

**Comment by hrydgard** [comment]
Thanks, added to list.

**Comment by Gamemulatorer** [comment]
Burnout Dominator sun flares is glitchy using the recently build PPSSPP.
![Screenshot_20230204_194908_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/216766183-45cd145f-818a-44f9-b7cc-6b7b57da434f.jpg)

**Comment by hrydgard** [comment]
Yeah, I'll have to take a look at those again.

**Comment by Gamemulatorer** [comment]
Resistance Retribution

***Software***
![Screenshot_20230318_065453_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/226066592-2eb66d85-f3d5-4e9f-b620-e4e9d27ffa42.jpg)

***Vulkan/OpenGL***
![Screenshot_20230318_065718_2f85358b2198d26f8aca533d68bee793](https://user-images.githubusercontent.com/37603562/226066618-36e670f0-d710-4bd8-88ac-ba2e16d4c2e6.jpg)

[UCES01184.ppdmp.zip](https://github.com/hrydgard/ppsspp/files/11006869/UCES01184.ppdmp.zip)

**Comment by hrydgard** [comment]
Alright, by implementing a new Z-only rasterizer, I have confirmed the theories above, that rasterizing depth-only can be made really really fast. I haven't found a single case yet in the affected games where alpha testing would have a noticeable effect on the lens flare blocking. But of course, it's possible we'll find one.

**Comment by hrydgard** [comment]
OK, with the new stuff in, I think we can almost call this fixed. Still need to confirm Resistance, and Split/Second I'm not sure about either.

**Comment by 71knight** [comment]
> OK, with the new stuff in, I think we can almost call this fixed. Still need to confirm Resistance, and Split/Second I'm not sure about either.

Okay I try it with this newest fix.  Sad to say with default lens flare settings ( skip GPU readbacks-no / occlusion- low ) it is still the same problem!  However with lens flare occlusion always on things are a little different.... some solid objects will block the sun while other solid objects exhibit the old behavior of letting the sun shine through.  When the sun is shining through these objects the screen is flickering.

**Comment by hrydgard** [comment]
Hm, Resistance: Retribution is on the list of games that should use the depth raster feature, so "low" should be almost the same as default or always. What's the game ID of your copy?

**Comment by 71knight** [comment]
> Hm, Resistance: Retribution is on the list of games that should use the depth raster feature, so "low" should be almost the same as default or always. What's the game ID of your copy?

Split second velocity - ULUS10513

**Comment by hrydgard** [comment]
Oh apologies, I misread your post, thought you were talking about Resistance. Yes, split second doesn't quite seem to be fixed, will need to look into it more.
